rtic-monotonics/src/stm32.rs

```diff
@@ -0,0 +1,216 @@
+pub use fugit::{self, ExtU64};
+use pac::TIM2;
+use stm32_metapac as pac;
+
+const TIMER_HZ: u32 = 1_000_000;
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __internal_create_stm32_timer_interrupt {
+    ($mono_timer:ident, $timer:ident, $timer_token:ident) => {{
+        #[no_mangle]
+        #[allow(non_snake_case)]
+        unsafe extern "C" fn $timer() {
+            $crate::stm32::$mono_timer::__tq().on_monotonic_interrupt();
+        }
+
+        pub struct $timer_token;
+
+        unsafe impl $crate::InterruptToken<$crate::stm32::$mono_timer> for $timer_token {}
+
+        $timer_token
+    }};
+}
+
+/// Register the TIM2 interrupt for the monotonic.
+#[macro_export]
+macro_rules! create_stm32_tim2_monotonic_token {
+    () => {{
+        $crate::__internal_create_stm32_timer_interrupt!(Tim2, TIM2, Tim2Token)
+    }};
+}
+
+macro_rules! make_timer {
+    ($mono_name:ident, $timer:ident, $overflow:ident, $tq:ident$(, doc: ($($doc:tt)*))?) => {
+        /// Monotonic timer queue implementation.
+        $(
+            #[cfg_attr(docsrs, doc(cfg($($doc)*)))]
+        )?
+
+        pub struct $mono_name;
+
+        static $overflow: AtomicU64 = AtomicU64::new(0);
+        static $tq: TimerQueue<$mono_name> = TimerQueue::new();
+
+        impl $mono_name {
+            /// Start monotonic timer. Must be called only once.
+            /// `tim_clock_hz` shows to which frequency `TIMx` clock source is configured.
+            pub fn start(tim_clock_hz: u32, _interrupt_token: impl crate::InterruptToken<Self>) {
+                pac::RCC.apbenr1().modify(|r| r.set_tim2en(true));
+                pac::RCC.apbrstr1().modify(|r| r.set_tim2rst(true));
+                pac::RCC.apbrstr1().modify(|r| r.set_tim2rst(false));
+
+                $timer.cr1().modify(|r| r.set_cen(false));
+
+                let psc = tim_clock_hz / TIMER_HZ - 1;
+                $timer.psc().write(|r| r.set_psc(psc as u16));
+
+                // Enable update event interrupt.
+                $timer.dier().modify(|r| r.set_uie(true));
+
+                // Trigger an update event to load the prescaler value to the clock.
+                $timer.egr().write(|r| r.set_ug(true));
+
+                // The above line raises an update event which will indicate that the timer is already finished.
+                // Since this is not the case, it should be cleared.
+                $timer.sr().modify(|r| r.set_uif(false));
+
+                // Start the counter.
+                $timer.cr1().modify(|r| {
+                    r.set_cen(true);
+                });
+
+                $tq.initialize(Self {});
+
+                // SAFETY: We take full ownership of the peripheral and interrupt vector,
+                // plus we are not using any external shared resources so we won't impact
+                // basepri/source masking based critical sections.
+                unsafe {
+                    //crate::set_monotonic_prio(pac::NVIC_PRIO_BITS, pac::Interrupt::$timer);
```

Comment:
@Dirbaio maybe you can tell me how to get `NVIC_PRIO_BITS` from `stm32-metapack`. It seems it is gated with `rt` feature flag, but when this was enabled then I got conflict with Interrupts.

Reply:
hm I think I made it gated by `rt` because that's what other PACs do. It's true that makes stm32-metapac generate the interrupt vectors etc which will conflict if your'e also using a `stm32-rs` PAC in the same project. 

I'm not sure if there's any downsides to always exposing it regardless of the `rt` feature, I don't think so? PRs welcome to always expose it.

Alternatively you can get it from [here](https://docs.embassy.dev/stm32-metapac/git/stm32wle4cb/metadata/struct.Metadata.html#structfield.nvic_priority_bits) but the metadata is intended to be used from `build.rs`, so maybe it's not what you're looking for.

Reply:
I couldn't figure out if it is possible to use `stm32-metapac` with `rt` in `rtic-monotonics` and at the same time import `embassy-stm32` to your main app without getting interrupt vectors conflict

```toml
embassy-stm32 = { git = "https://github.com/embassy-rs/embassy.git", features = ["defmt", "stm32g081kb", "nightly", "unstable-traits", "unstable-pac", "exti"] }
rtic-monotonics = { path = "../rtic/rtic-monotonics", features = ["stm32g081kb"]}
```
So for now I am gonna use `METADATA.nvic_priority_bits.unwrap()` as you suggested.